### ISSUE TITLE
Connected instance shall be invalidated once disconnected

### ISSUE THREAD

**Issue by postlund** [postlund/pyatv#1415]
### Short feature/function description

After disconnecting, it's still possible to call functions in the interface and it will most certainly result in an error, but unrelated to being disconnected. I would prefer if a more appropriate exception is raised when calling methods on a disconnected object.

It's hard to handle every single case here since a user might save a handle to anything, but as much as possible shall be guarded. That means this should raise an exception:

```python
atv = await connect(...)
rc = atv.remote_control
atv.disconnect()
await rc.stop()  # raises exception
```

### What needs to be done?

I'm not sure if I want to add a `DisconnectedError` or just re-use `InvalidStateError` here. But one of them will do.

Most functions are easy to guard in `core.FacadeAppleTV` by just using an if-statement. The relayers are however a bit tricky, so I think I need to extend `Relayer` with a method that invalidate it, making `Relayer.relay` raise an exception. That would align more with `InvalidStateError` rather than `DisconnectedError` as the relayer has no notion of connection state.

### Is this a breaking change?

No (it would be an erroneous use of the library if relying on current behavior)

### Anything else worth knowing?

No

**Comment by postlund** [comment]
So, I started poking around with this and I took another route to make the implementation more generic. The problem with involving the relayer and treating methods in `interface.AppleTV` as special makes the code clumsy.

What I did instead was experimenting with a concept called "shielding". An object can be "shielded" and methods within that object/class can be "guarded". If a shielded object is moved to blocking state, all guarded methods will automatically block all calls by raising an exception. A simple example looks like this:

```python
from pyatv.support.shield import block, guard, shield

class MyClass:
    @guard
    def guarded(self):
        print("do something")


obj = shield(MyClass())
obj.guarded()
block(obj)
obj.guarded()
```

The first call to `guarded` will print the text as expected, but the second call (after `block`) will raise an exception:

```raw
do something
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../pyatv/pyatv/support/shield.py", line 70, in _guard_method
    raise BlockedStateError(f"{func.__name__} is blocked")
pyatv.exceptions.BlockedStateError: guarded is blocked
```

All methods belonging to the public interface will be guarded and a disconnect triggers blocking. It's a bit tricky to get everything right, but I'm almost done.
